File: haskell/ch1/Ex114.hs
-- :cd C:\Users\Yohan\IdeaProjects\sicp\haskell\ch1
-- :load "ExXXX.hs"
-- Run with count_change 100 (should equal 292)

module Ex114 where

biggest_denomiation 1 = 1
biggest_denomiation 2 = 5
biggest_denomiation 3 = 10
biggest_denomiation 4 = 25
biggest_denomiation 5 = 50

cc 0 _ = 1
cc _ 0 = 0
cc amount _ | amount < 0 = 0
cc amount kinds_of_coins = cc amount (kinds_of_coins - 1) + cc (amount - (biggest_denomiation kinds_of_coins)) kinds_of_coins

count_change amount = cc amount 5

File: erlang/ch1/ex114.erl
%%%-------------------------------------------------------------------
%%% @author Yohan Lee
%%% @copyright (C) 2019
%%% @doc
%%% cd("c:/Users/Yohan/IdeaProjects/sicp/erlang/ch1").
%%% compile with > c("exXXX.erl").
%%% if already compiled, load with > l(exXXX).
%%% run with > exXXX:main().
%%%
%%% @end
%%%-------------------------------------------------------------------
-module(ex114).
-author("Yohan").

%% API
-export([main/0]).

main() ->
  io:fwrite("~w\n", [count_change(1)]),
  io:fwrite("~w\n", [count_change(2)]),
  io:fwrite("~w\n", [count_change(3)]),
  io:fwrite("~w\n", [count_change(11)]),
  io:fwrite("~w\n", [count_change(100)]).


biggest_denom(1) -> 1;
biggest_denom(2) -> 5;
biggest_denom(3) -> 10;
biggest_denom(4) -> 25;
biggest_denom(5) -> 50.


cc(0, _) -> 1;
cc(_, 0) -> 0;
cc(Amount, _) when Amount < 0 -> 0;
cc(Amount, KindsOfCoins) ->
  cc(Amount, KindsOfCoins - 1) + cc(Amount - biggest_denom(KindsOfCoins), KindsOfCoins).


count_change(Amount) -> cc(Amount, 5).

%% count_change(100) == 292
